src/main/ApplicationImpl.cpp

```diff
@@ -902,4 +902,9 @@ ApplicationImpl::getLedgerTxnRoot()
     return mConfig.MODE_USES_IN_MEMORY_LEDGER ? *mNeverCommittingLedgerTxn
                                               : *mLedgerTxnRoot;
 }
+
+void
+ApplicationImpl::actBeforeDBSchemaUpgrade()
```

Comment:
I'm not sure this is the cleanest approach. It has the annoying overhead of needing to create a derived class for each variation you want. Perhaps it would be better to add a test-only setting (eg. one that can't be specified in the configuration file) to `Config` called something like `UPGRADE_TO_SCHEMA_VERSION` (default value is the current schema version). Then instead of `ApplicationImpl` calling `mDatabase->upgradeToCurrentSchema` you can call `mDatabase->upgradeToSchema(mConfig.UPGRADE_TO_SCHEMA_VERSION)`. With this, you can just create an application at the desired schema version then upgrade it again later. Do you think this is better or worse?

Reply:
I considered a Config option at first, and I might be misunderstanding something but I don't think that it would have worked:  the application creation itself, when it creates a new database, ends up calling LedgerManagerImpl::startNewLedger(), which creates a root account and then closes the ledger, which in turn updates the database, and the database update code expects the database to be in the current schema already.  So we can't wait until after createTestApplication() returns before upgrading the schema version.

Furthermore, even if a Config option would have worked, I'd still like what I did better, for several reasons:

- Because the derived class constructor can take whatever parameters it wants (because createTestApplication() and Application::create() take variadic parameter packs), and in particular can take lambdas for one or more parameters, we don't need to create a new derived class for each different function that we might want to inject into the code path between database creation and schema upgrade; we can just pass a different lambda for each different function, as illustrated in the code I've already written using the SchemaUpgradeTestApplication derived class, and each different function could take different sets of (and numbers of) parameters.  We only need to create a new derived class if and when we reuse the mechanism to add test-code-injection functionality to _other_ code paths, and that would also be true of a Config option -- if we wanted to add test-code-injection functionality to more code paths in the future, we'd have to add new Config options for those.

- The code that handles the weird C++ variadic-parameter syntax only has to be written once (and is written in this check-in) -- clients (as illustrated by SchemaUpgradeTestApplication) don't have to use that syntax; they simply write their constructors with however many parameters they want, and pass a matching list of parameters to createTestApplication().

- With this code, since the variadic parameters propagate to Application::create(), any other future derived classes of Application itself, or future users of Application::create with different template parameters, would also get the option of having arbitrary-length-and-type parameter lists passed to their constructors as part of Application::create().

- With this code, each call to createTestApplication() can store a different set of parameters in its template parameter class (derived from TestApplication or Application), which would then be accessible to the callback from the injection point, so that that same callback would work, with the same list of parameter types (including an empty list, as in this case), for different invocations of createTestApplication() using the same derived class.  To accomplish that with a Config option without deriving from TestApplication, I think that we'd have to do something such as including potential parameters to the injection function within TestApplication itself, and setting them manually in the client before calling createTestApplication(), which would both litter the base TestApplication with members having to do with injections that were only used by certain individual tests, and require clients to make multiple ->setParameter() calls to set those parameters prior to calling createTestApplication(), rather than simply using them as parameters to createTestApplication() itself.

In short, I don't think a Config option works, because of the code called within Application::create() itself that relies on the schema already having been upgraded, but even if it did work, I think that what I did would provide a lot more flexibility for clean future test-code injections, so I'd still prefer it.  But of course I'm open to counter-arguments!